# NETFLIX EDA PROJECT - SQL

Netflix, a leading global streaming platform, possesses a dataset containing information about its shows. However, the dataset requires cleaning and analysis to derive valuable insights for business decision-making on the Netflix dataset to help the company gain insights into their content offerings.

**STEP 1 - Importing all the required libraries**

In [1]:
import numpy as np
import pandas as pd
import sqlite3

import warnings
warnings.filterwarnings('ignore')

**STEP 2 - Loading the data into jupyter notebook and forming connection with Sqlite with netflix database**

In [3]:
df = pd.read_csv("C:\\Users\\SAUMIL SHAH\\Desktop\\Netflix.csv", encoding= 'unicode_escape')

conn = sqlite3.connect("Netflix_db")

**STEP 3 - Giving the table name as "netflix_data" and creating column schema**

In [9]:
Table_name = 'netflix_data1'
Column_schema ='show_id INTEGER,show_type TEXT, title TEXT, director TEXT,country TEXT,date_added INTEGER,release_year INTEGER,rating TEXT,duration TEXT,listed_in TEXT'

**STEP 4 - Creating the table query**

In [5]:
Create_Table_Query = f"CREATE TABLE {Table_name} ({Column_schema})"
conn.execute(Create_Table_Query)

**STEP 5 - Writing Pandas Dataframe(df) to SQL database table**

In [10]:
df.to_sql(Table_name, conn, if_exists='append', index=False)

8766

**Step 6 - Checking the datatypes**

In [58]:
df.dtypes

show_id         object
show_type       object
title           object
director        object
country         object
date_added      object
release_year     int64
rating          object
duration        object
listed_in       object
dtype: object

**Step 7 - As we can se date_added datatype is object therefore we need to convert it to datetime format**

In [89]:
df['date_added'] = pd.to_datetime(df['date_added'])
df.dtypes

show_id                 object
show_type               object
title                   object
director                object
country                 object
date_added      datetime64[ns]
release_year             int64
rating                  object
duration                object
listed_in               object
dtype: object

# Segment 1: Database - Tables, Columns, Relationships

**A. Identify the tables in the dataset and their respective columns**

In [90]:
seg1_a = pd.read_sql("""SELECT *
                        FROM netflix_data1;""",conn)
display(seg1_a)

show_id show_type                             title         director  \
0         s1     Movie              Dick Johnson Is Dead  Kirsten Johnson   
1         s3   TV Show                         Ganglands  Julien Leclercq   
2         s6   TV Show                     Midnight Mass    Mike Flanagan   
3        s14     Movie  Confessions of an Invisible Girl    Bruno Garotti   
4         s8     Movie                           Sankofa     Haile Gerima   
...      ...       ...                               ...              ...   
8761   s8797   TV Show                        Yunus Emre        Not Given   
8762   s8798   TV Show                         Zak Storm        Not Given   
8763   s8801   TV Show                Zindagi Gulzar Hai        Not Given   
8764   s8784   TV Show                              Yoko        Not Given   
8765   s8786   TV Show                               YOM        Not Given   

            country  date_added  release_year rating   duration  \
0     United States  25-09-2021          2020  PG-13     90 min   
1            France  24-09-2021          2021  TV-MA   1 Season   
2     United States  24-09-2021          2021  TV-MA   1 Season   
3            Brazil  22-09-2021          2021  TV-PG     91 min   
4     United States  24-09-2021          1993  TV-MA    125 min   
...             ...         ...           ...    ...        ...   
8761         Turkey  17-01-2017          2016  TV-PG  2 Seasons   
8762  United States  13-09-2018          2016  TV-Y7  3 Seasons   
8763       Pakistan  15-12-2016          2012  TV-PG   1 Season   
8764       Pakistan  23-06-2018          2016   TV-Y   1 Season   
8765       Pakistan  06-07-2018          2016  TV-Y7   1 Season   

                                              listed_in  
0                                         Documentaries  
1     Crime TV Shows, International TV Shows, TV Act...  
2                    TV Dramas, TV Horror, TV Mysteries  
3                    Children & Family Movies, Comedies  
4      Dramas, Independent Movies, International Movies  
...                                                 ...  
8761                  International TV Shows, TV Dramas  
8762                                           Kids' TV  
8763  International TV Shows, Romantic TV Shows, TV ...  
8764                                           Kids' TV  
8765                                           Kids' TV  

[8766 rows x 10 columns]

From above dataset we can understand that there are total 10 columns and 1 table

**B. Determine the number of rows in each table within the schema**

In [91]:
seg1_b = pd.read_sql("""
SELECT count(*) as rows
FROM 
netflix_data1;""",conn)
display(seg1_b)

rows
0  8766

There are total 8766 number of rows in this table within the schema

**C. Identify and handle any missing values in the dataset**

In [92]:
seg1_c = pd.read_sql("""
SELECT count(*) 
FROM 
netflix_data1
Where show_id is null OR title is null OR show_type is null OR director is null OR 
country is null OR date_added is null OR release_year is null OR rating is null OR 
duration is null OR listed_in is null;
""",conn)
display(seg1_c)

count(*)
0         0

As we can see there are no missing values in the dataset

# Segment 2: Content Analysis

**A. Analyse the distribution of content types (movies vs. TV shows) in the dataset**

In [93]:
seg2_a = pd.read_sql("""
SELECT show_type, count(*) as distribution
FROM 
netflix_data1
group by show_type;
""",conn)
display(seg2_a)

show_type  distribution
0     Movie          6108
1   TV Show          2658

As per above data we can analyse that there are more number of movies which is 6108 in compare to tv show which is 2658 respectively on netflix.

**B. Determine the top 10 countries with the highest number of productions on Netflix**

In [94]:
seg2_b = pd.read_sql("""
SELECT Country, count(*) as no_of_productions
FROM 
netflix_data1
GROUP BY country
ORDER BY no_of_productions DESC
limit 10;
""",conn)
display(seg2_b)

country  no_of_productions
0   United States               3234
1           India               1053
2  United Kingdom                638
3        Pakistan                419
4       Not Given                287
5          Canada                271
6           Japan                259
7     South Korea                214
8          France                213
9           Spain                182

As per above data top 10 countries with the highest number of productions on netflix are US, india, UK, Pakistan, not given, Canada, Japan, South Korea, France, Spain.

**C. Investigate the trend of content additions over the years**

In [95]:
seg2_c = pd.read_sql("""
SELECT release_year as year, count(*) as Content_count
FROM 
netflix_data1
GROUP BY release_year
ORDER BY release_year
""",conn)
display(seg2_c)

year  Content_count
0   1925              1
1   1942              2
2   1943              3
3   1944              3
4   1945              4
..   ...            ...
69  2017           1029
70  2018           1140
71  2019           1024
72  2020            952
73  2021            592

[74 rows x 2 columns]

As per above data it can be seen that year by year the releasing of content are increasing.

**D. Analyse the relationship between content duration and release year**

In [96]:
seg2_d = pd.read_sql("""
SELECT release_year as year, 
        CASE
          WHEN show_type = 'Movie' THEN AVG(CAST(SUBSTR(duration, 1, INSTR(duration, ' min') - 1) AS INT))
          WHEN show_type = 'TV Show' THEN AVG(CAST(SUBSTR(duration, 1, INSTR(duration, ' Season') - 1) AS INT)*5 * 45)
          ELSE 0
        END AS content_duration
FROM 
netflix_data1
WHERE show_type = 'Movie' OR (show_type = 'TV Show' AND INSTR(duration, ' Season') > 0)
GROUP BY release_year
ORDER BY year;
""",conn)
display(seg2_d)

year  content_duration
0   1925        225.000000
1   1942         35.000000
2   1943         62.666667
3   1944         52.000000
4   1945         38.500000
..   ...               ...
69  2017        103.644315
70  2018        124.736842
71  2019         57.238281
72  2020         50.038866
73  2021        205.616554

[74 rows x 2 columns]

In this query Movie is in min and TV Shows are in seasons there for we need to calculate avg duration for both simultaneously I used nested case statement in which for "Movie used SUBSTR(duration, 1, INSTR(duration, 'min') - 1) to extracts the substring from the duration column starting from the first character and ending before 'min' and in next line for "TV Show used SUBSTR(duration, 1, INSTR(duration, ' Season') - 1) to extracts the substring from the duration column starting from the first character and ending before ' Season'. This gives us the number of seasons. Therefore CAST(... AS INT)5 45 converts the extracted substring to an integer and multiplies it by the assumed value of 5 episodes in each season which are of 45 min duration

**E. Identify the directors with the most content on Netflix**

In [97]:
seg2_e = pd.read_sql("""
SELECT Director, count(*) as most_content
FROM 
netflix_data1
GROUP BY director
ORDER BY most_content DESC
Limit 20
""",conn)
display(seg2_e)

director  most_content
0                      Not Given          2584
1                  Rajiv Chilaka            20
2        Ra?§l Campos, Jan Suter            18
3            Alastair Fothergill            18
4                    Suhas Kadav            16
5                   Marcus Raboy            16
6                      Jay Karas            14
7            Cathy Garcia-Molina            13
8                Youssef Chahine            12
9                Martin Scorsese            12
10                   Jay Chapman            12
11              Steven Spielberg            11
12  Mark Thornton, Todd Kauffman            10
13              Don Michael Paul            10
14                  David Dhawan             9
15              YŽñlmaz ErdoŽ?an             8
16                   Troy Miller             8
17               Shannon Hartman             8
18                   Ryan Polito             8
19              Robert Rodriguez             8

As per above data it can be seen that director is not mentioned for most of the content

# Segment 3: Genre and Category Analysis

**A. Determine the unique genres and categories present in the dataset**

In [98]:
seg3_a = pd.read_sql("""
SELECT distinct(listed_in) as categories
FROM 
netflix_data1
""",conn)
display(seg3_a)

categories
0                                        Documentaries
1    Crime TV Shows, International TV Shows, TV Act...
2                   TV Dramas, TV Horror, TV Mysteries
3                   Children & Family Movies, Comedies
4     Dramas, Independent Movies, International Movies
..                                                 ...
508         Classic & Cult TV, TV Horror, TV Mysteries
509                        Crime TV Shows, TV Comedies
510           Classic & Cult TV, Kids' TV, TV Comedies
511             Classic & Cult TV, TV Sci-Fi & Fantasy
512       Classic & Cult TV, Crime TV Shows, TV Dramas

[513 rows x 1 columns]

Above are some unique genres and catergories in netflix dataset.

**B. Calculate the percentage of movies and TV shows in each genre**

In [99]:
seg3_b = pd.read_sql("""
SELECT listed_in as genre,
(SUM(CASE WHEN show_type = 'Movie' THEN 1 ELSE 0 END) * 100.0) / COUNT(*) as movie_percentage,
(SUM(CASE WHEN show_type = 'TV Show' THEN 1 ELSE 0 END) * 100.0) / COUNT(*) as tv_show_percentage
FROM 
netflix_data1
GROUP BY listed_in;
""",conn)
display(seg3_b)

genre  movie_percentage  \
0                                   Action & Adventure             100.0   
1                   Action & Adventure, Anime Features             100.0   
2    Action & Adventure, Anime Features, Children &...             100.0   
3    Action & Adventure, Anime Features, Classic Mo...             100.0   
4    Action & Adventure, Anime Features, Horror Movies             100.0   
..                                                 ...               ...   
508             TV Horror, TV Mysteries, Teen TV Shows               0.0   
509                           TV Horror, Teen TV Shows               0.0   
510                  TV Sci-Fi & Fantasy, TV Thrillers               0.0   
511                                           TV Shows               0.0   
512                                          Thrillers             100.0   

     tv_show_percentage  
0                   0.0  
1                   0.0  
2                   0.0  
3                   0.0  
4                   0.0  
..                  ...  
508               100.0  
509               100.0  
510               100.0  
511               100.0  
512                 0.0  

[513 rows x 3 columns]

As per above data we can see the percentage of movie and tv show as per genre

**C. Identify the most popular genres/categories based on the number of productions**

In [100]:
seg3_C = pd.read_sql("""
SELECT listed_in as Popular_Categories,Count(show_type) as no_of_production
FROM 
netflix_data1
GROUP BY listed_in
ORDER BY no_of_production DESC
Limit 10;
""",conn)
display(seg3_C)

Popular_Categories  no_of_production
0                      Dramas, International Movies               361
1                                     Documentaries               359
2                                   Stand-Up Comedy               334
3            Comedies, Dramas, International Movies               274
4  Dramas, Independent Movies, International Movies               251
5                                          Kids' TV               219
6                          Children & Family Movies               215
7                Children & Family Movies, Comedies               201
8               Documentaries, International Movies               186
9     Dramas, International Movies, Romantic Movies               180

As per above data most popular categories on netflix is "Dramas, International Movies" with 361 number of productions.

**D. Calculate the cumulative sum of content duration within each genre**

In [142]:
seg3_d = pd.read_sql("""
Select listed_in AS genre,duration,
Sum(duration) Over(PARTITION BY listed_in ORDER BY date_added) as cumulative_sum
from netflix_data1
Where show_type != 'TV Show';    
""",conn)
display(seg3_d)

genre duration  cumulative_sum
0     Action & Adventure  101 min           101.0
1     Action & Adventure   88 min          1056.0
2     Action & Adventure   97 min          1056.0
3     Action & Adventure  111 min          1056.0
4     Action & Adventure  137 min          1056.0
...                  ...      ...             ...
6103           Thrillers   97 min          6106.0
6104           Thrillers  100 min          6206.0
6105           Thrillers   99 min          6305.0
6106           Thrillers  108 min          6413.0
6107           Thrillers   84 min          6497.0

[6108 rows x 3 columns]

As per above data we can see cumulative sum of content duration with respect to each genre.

# Segment 4: Release Date Analysis

**A. Determine the distribution of content releases by month and year**

In [108]:
seg4_a= pd.read_sql("""
select strftime('%Y', date_added) AS release_year,
strftime('%m', date_added) AS release_month,
Count(*) AS release_count
From netflix_data1
Group by release_year, release_month
Order by release_count desc;
""", conn)

display(seg4_a)

release_year release_month  release_count
0          None          None           1140
1          None          None           1029
2          None          None           1024
3          None          None            952
4          None          None            899
..          ...           ...            ...
69         None          None              1
70         None          None              1
71         None          None              1
72         None          None              1
73         None          None              1

[74 rows x 3 columns]

**B. Analyse the seasonal patterns in content releases**

In [143]:
seg4_b = pd.read_sql("""
Select case
        WHEN strftime('%m', date_added) IN ('12', '01', '02') THEN 'Winter'
        WHEN strftime('%m', date_added) IN ('03', '04', '05') THEN 'Summer'
        WHEN strftime('%m', date_added) IN ('06', '07', '08') THEN 'Rainy'
        WHEN strftime('%m', date_added) IN ('09', '10', '11') THEN 'Autumn'
            ELSE 'Unknown'
       END AS season,
COUNT(*) AS release_count
From netflix_data1
Group by season
Order by release_count desc;
""",conn)
display(seg4_b)

season  release_count
0  Unknown           8766

**C. Identify the months and years with the highest number of releases**

In [145]:
seg4_c = pd.read_sql("""
Select strftime('%Y-%m', date_added) AS release_month_year,
Count(*) AS release_count
From netflix_data1
Group by release_month_year
ORDER BY release_count DESC;
""",conn)
display(seg4_c)

release_month_year  release_count
0               None           8766

# Segment 5: Rating Analysis

**A. Investigate the distribution of ratings across different genres**

In [115]:
seg5_a= pd.read_sql("""
SELECT listed_in as genre,rating, count(*) as COUNT
From netflix_data1
Group by genre,rating
ORDER BY COUNT DESC;
""", conn)

display(seg5_a)

genre rating  COUNT
0                                      Stand-Up Comedy  TV-MA    284
1                         Dramas, International Movies  TV-MA    153
2     Dramas, Independent Movies, International Movies  TV-MA    142
3               Comedies, Dramas, International Movies  TV-14    139
4                         Dramas, International Movies  TV-14    139
...                                                ...    ...    ...
1226     TV Dramas, TV Sci-Fi & Fantasy, Teen TV Shows  TV-14      1
1227                           TV Dramas, TV Thrillers  TV-14      1
1228            TV Horror, TV Mysteries, Teen TV Shows  TV-MA      1
1229                 TV Sci-Fi & Fantasy, TV Thrillers  TV-14      1
1230                                          TV Shows      R      1

[1231 rows x 3 columns]

As per above data it is seen that highes number of ratings is given to "stand up comedy"

**B. Analyse the relationship between ratings and content duration**

In [119]:
seg5_b= pd.read_sql("""
SELECT rating,
AVG(CASE
WHEN show_type = 'Movie' THEN CAST(SUBSTR(duration, 1, INSTR(duration, ' min') - 1) AS INT)
WHEN show_type = 'TV Show' THEN CAST(SUBSTR(duration, 1, INSTR(duration, ' Season') - 1) AS INT)*5 * 45
ELSE 0
END) AS content_duration
From netflix_data1
Group by rating
ORDER BY content_duration DESC;
""", conn)

display(seg5_b)

rating  content_duration
0      TV-Y7        292.534535
1       TV-Y        258.833333
2       TV-G        223.581818
3      TV-14        210.846977
4      TV-MA        196.660513
5      TV-PG        196.161440
6   TV-Y7-FV        132.000000
7      NC-17        125.000000
8      PG-13        108.256674
9          R        106.964736
10        NR        106.835443
11        UR        106.333333
12        PG         98.282230
13         G         90.268293

As per above data we can see relationship between ratings and their respective content duration and it can be seen that rating TV-Y7 containing highest number of duration.

# Segment 6: Co-occurrence Analysis

**A. Identify the most common pairs of genres/categories that occur together in content**

In [139]:
seg6_a= pd.read_sql("""
SELECT n1.listed_in AS genre1, 
       n2.listed_in AS genre2, 
       COUNT(*) AS pair_count
FROM netflix_data1 n1
JOIN netflix_data1 n2 
ON 
n1.show_id != n2.show_id -- To avoid counting pairs twice
AND 
(n1.listed_in < n2.listed_in -- To avoid counting same pairs in reverse order
GROUP BY genre1, genre2
ORDER BY pair_count DESC
LIMIT 10;
""", conn)

display(seg6_a)

genre1  \
0                           Documentaries   
1            Dramas, International Movies   
2            Dramas, International Movies   
3                         Stand-Up Comedy   
4                           Documentaries   
5                         Stand-Up Comedy   
6  Comedies, Dramas, International Movies   
7            Dramas, International Movies   
8  Comedies, Dramas, International Movies   
9                           Documentaries   

                                   genre2  pair_count  
0            Dramas, International Movies      129599  
1                           Documentaries      129599  
2                         Stand-Up Comedy      120574  
3            Dramas, International Movies      120574  
4                         Stand-Up Comedy      119906  
5                           Documentaries      119906  
6            Dramas, International Movies       98914  
7  Comedies, Dramas, International Movies       98914  
8                           Documentaries       98366  
9  Comedies, Dramas, International Movies       98366

By implimenting self join we are able to see most common pairs of genre 

**B. Analyse the relationship between genres/categories and content duration**

In [130]:
seg6_b= pd.read_sql("""
SELECT listed_in AS genre,
round(AVG(CASE
WHEN show_type = 'Movie' THEN CAST(SUBSTR(duration, 1, INSTR(duration, ' min') - 1) AS INT)
WHEN show_type = 'TV Show' THEN CAST(SUBSTR(duration, 1, INSTR(duration, ' Season') - 1) AS INT)*5 * 45
ELSE 0
END)) AS content_duration
From netflix_data1
Group by genre
ORDER BY content_duration DESC;
""", conn)

display(seg6_b)

genre  content_duration
0    Classic & Cult TV, TV Action & Adventure, TV H...            2025.0
1                       Classic & Cult TV, TV Comedies            1800.0
2    Crime TV Shows, TV Action & Adventure, TV Sci-...            1575.0
3    Classic & Cult TV, TV Action & Adventure, TV S...            1575.0
4     British TV Shows, Classic & Cult TV, TV Comedies            1463.0
..                                                 ...               ...
508                                             Movies              46.0
509   Children & Family Movies, Comedies, LGBTQ Movies              46.0
510   Action & Adventure, Documentaries, Sports Movies              40.0
511                      Anime Features, Documentaries              36.0
512          Comedies, Music & Musicals, Sports Movies              30.0

[513 rows x 2 columns]

As per above data we can see relationship between genre and content duration.

# Segment 7: International Expansion Analysis

**A. Identify the countries where Netflix has expanded its content offerings**

In [131]:
seg7_a= pd.read_sql("""
SELECT Distinct country
from 
netflix_data1;
""", conn)

display(seg7_a)

country
0    United States
1           France
2           Brazil
3   United Kingdom
4            India
..             ...
81         Senegal
82         Belarus
83     Puerto Rico
84          Cyprus
85         Ukraine

[86 rows x 1 columns]

As per above data it can be seen Netflix has content offerings in almost 86 countries.

**B. Analyse the distribution of content types in different countries**

In [135]:
seg7_b= pd.read_sql("""
SELECT country,show_type, count(*) as COUNT
from 
netflix_data1
GROUP BY country,show_type
ORDER BY COUNT DESC;
""", conn)
display(seg7_b)

country show_type  COUNT
0           United States     Movie   2389
1                   India     Movie    972
2           United States   TV Show    845
3          United Kingdom     Movie    387
4                Pakistan   TV Show    348
..                    ...       ...    ...
133           Switzerland   TV Show      1
134  United Arab Emirates   TV Show      1
135               Uruguay   TV Show      1
136          West Germany     Movie      1
137              Zimbabwe     Movie      1

[138 rows x 3 columns]

As per above data we can analyse that US have highest number of movies as content types

**C. Investigate the relationship between content duration and country of production**

In [136]:
seg7_c= pd.read_sql("""
SELECT country,
round(AVG(CASE
WHEN show_type = 'Movie' THEN CAST(SUBSTR(duration, 1, INSTR(duration, ' min') - 1) AS INT)
WHEN show_type = 'TV Show' THEN CAST(SUBSTR(duration, 1, INSTR(duration, ' Season') - 1) AS INT)*5 * 45
ELSE 0
END)) AS content_duration
From netflix_data1
Group by country
ORDER BY content_duration DESC;
""", conn)

display(seg7_c)

country  content_duration
0     Ukraine             450.0
1     Belarus             450.0
2     Croatia             304.0
3     Denmark             287.0
4      Russia             276.0
..        ...               ...
81   Slovenia              84.0
82    Georgia              72.0
83  Guatemala              69.0
84      Syria              52.0
85    Namibia              29.0

[86 rows x 2 columns]

As per above data we can see that Ukraine and belarus have highest content duration

# Segment 8: Recommendations for Content Strategy

**Based on the analysis, provide recommendations for the types of content Netflix should focus on producing and Identify potential areas for expansion and growth based on the analysis of the dataset**

Based on the analysis conducted in the previous segments, we can provide recommendations for the types of content Netflix should focus on producing and identify potential areas for expansion and growth. Here are some suggestions:

1)Content Types: Analyzing the distribution of content types (movies and TV shows) in the dataset. Identify which type is more popular and has higher demand among viewers. Based on this analysis, we can prioritize the production of the more popular content type.

2)Genres and Categories: Analyzing the distribution of genres and categories in the dataset. Identify the most popular genres and categories based on the number of productions. Focus on producing content in these popular genres and categories to cater to viewer preferences.

3)Content Duration: Analyzing the relationship between content duration and viewer engagement. Identify the ideal content duration that keeps viewers engaged and interested. Based on this analysis, produce content with durations that align with viewer preferences and attention spans.

4)Ratings and Viewer Feedback: Consider the distribution of ratings across different genres and analyze the relationship between ratings and content duration. Pay attention to genres with higher ratings and positive viewer feedback. Use this information to guide the production of high-quality content that resonates with viewers.

5)Co-occurrence Analysis:This analysis can provide insights into potential cross-genre content combinations that have a higher chance of success.

6)International Expansion: Analyze the countries with the highest number of productions on Netflix. Consider expanding content offerings in these countries or explore opportunities to collaborate with local production industries to tap into new markets.

7)Seasonal Patterns: Investigate the seasonal patterns in content releases. Identify periods with higher content additions and plan content strategies accordingly to align with viewers' preferences during specific seasons.

8)Emerging Trends: Stay updated with emerging trends in the entertainment industry. Monitor viewership patterns, popular genres, and content formats to adapt and produce content that caters to evolving viewer preferences.

These recommendations and insights can help shape Netflix's content strategy, prioritize production efforts, and identify potential areas for expansion and growth